plugins/scaffolder-backend/src/service/rules.ts

```diff
@@ -42,4 +48,89 @@ export const hasTag = createScaffolderPermissionRule({
   toQuery: () => ({}),
 });
 
+export const createActionPermissionRule = makeCreatePermissionRule<
+  {
+    action: string;
+    input: JsonObject | undefined;
+  },
+  {},
+  typeof RESOURCE_TYPE_SCAFFOLDER_ACTION
+>();
+
+export const hasActionId = createActionPermissionRule({
+  name: 'HAS_ACTION_ID',
+  resourceType: RESOURCE_TYPE_SCAFFOLDER_ACTION,
+  description: `Match actions with the given actionId`,
+  paramsSchema: z.object({
+    actionId: z.string().describe('Name of the actionId to match on'),
+  }),
+  apply: (resource, { actionId }) => {
+    return resource.action === actionId;
+  },
+  toQuery: () => ({}),
+});
+
+export const hasProperty = buildHasProperty({
+  name: 'HAS_PROPERTY',
+  valueSchema: z.union([z.string(), z.number(), z.boolean(), z.null()]),
+  validateProperty: false,
+});
+
+export const hasBooleanProperty = buildHasProperty({
+  name: 'HAS_BOOLEAN_PROPERTY',
+  valueSchema: z.boolean(),
+});
+export const hasNumberProperty = buildHasProperty({
+  name: 'HAS_NUMBER_PROPERTY',
+  valueSchema: z.number(),
+});
+export const hasStringProperty = buildHasProperty({
+  name: 'HAS_STRING_PROPERTY',
+  valueSchema: z.string(),
+});
+
+function buildHasProperty<Schema extends z.ZodType<JsonPrimitive>>({
+  name,
+  valueSchema,
+  validateProperty = true,
+}: {
+  name: string;
+  valueSchema: Schema;
+  validateProperty?: boolean;
+}) {
+  return createActionPermissionRule({
+    name,
+    description: `Allow actions with the specified property`,
+    resourceType: RESOURCE_TYPE_SCAFFOLDER_ACTION,
+    paramsSchema: z.object({
+      key: z
+        .string()
+        .describe(`Property within the action parameters to match on`),
+      value: valueSchema.describe(`Value of the given property to match on`),
+    }) as unknown as z.ZodType<{ key: string; value?: z.infer<Schema> }>,
```

Comment:
Is this needed?

Reply:
it is! We would like to have more eyes on. Our goal was to make this function as generic as possible, to reduce the amount of duplicated code, we struggled a lot with typings and eventually we ended up with this ugly casting in order to move on.